File: java/src/test/java/it/proconsole/learning/shortestpath/parallelization/model/GraphTest.java
package it.proconsole.learning.shortestpath.parallelization.model;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

abstract class GraphTest {
  static final int VERTICES = 5;
  static final int X = 1;
  static final int Y = 2;
  static final int W = 3;
  static final int Z = 4;
  static final int A_WEIGHT = 9;
  private static final int A_NEGATIVE_WEIGHT = -4;

  abstract Graph graph();

  @Test
  void addBidirectionalEdge() {
    graph().addBidirectionalEdge(X, Y, A_WEIGHT);

    assertEquals(A_WEIGHT, graph().getCost(X, Y));
    assertEquals(A_WEIGHT, graph().getCost(Y, X));
  }

  @Test
  void addEdgeAndGetCost() {
    graph().addEdge(X, Y, A_WEIGHT);

    assertEquals(A_WEIGHT, graph().getCost(X, Y));

    graph().addEdge(X, Y, A_NEGATIVE_WEIGHT);

    assertEquals(A_NEGATIVE_WEIGHT, graph().getCost(X, Y));
  }

  @Test
  void hasNegativeEdges() {
    graph().addBidirectionalEdge(X, Y, A_WEIGHT);

    assertFalse(graph().hasNegativeEdges());

    graph().addBidirectionalEdge(X, Y, A_NEGATIVE_WEIGHT);

    assertTrue(graph().hasNegativeEdges());
  }

  @Test
  void haveConnection() {
    graph().addEdge(X, Y, A_WEIGHT);

    assertTrue(graph().haveConnection(X, Y));

    graph().addEdge(X, Y, Graph.ZERO_WEIGHT);

    assertFalse(graph().haveConnection(X, Y));
  }

  @Test
  void neighboursOf() {
    graph().addEdge(X, Y, A_WEIGHT);
    graph().addEdge(X, Z, A_WEIGHT);
    graph().addEdge(Y, W, A_WEIGHT);

    var current = graph().neighboursOf(X);

    var edges = List.of(new Edge(Y, A_WEIGHT), new Edge(Z, A_WEIGHT));
    assertEquals(edges, current);
  }

  @Test
  void removeEdge() {
    graph().addBidirectionalEdge(X, Y, A_WEIGHT);

    assertNotEquals(Graph.ZERO_WEIGHT, graph().getCost(X, Y));
    assertNotEquals(Graph.ZERO_WEIGHT, graph().getCost(Y, X));

    graph().removeEdge(X, Y);

    assertEquals(Graph.ZERO_WEIGHT, graph().getCost(X, Y));
    assertEquals(Graph.ZERO_WEIGHT, graph().getCost(Y, X));
  }

  @Test
  void vertices() {
    assertEquals(VERTICES, graph().vertices());
  }
}

File: ts/src/graph/graph.test.ts
import { Edge, matrixGraph } from './graph';

describe('matrixGraph', () => {
	const VERTICES = 5;
	const X = 1;
	const Y = 2;
	const W = 3;
	const Z = 4;
	const A_WEIGHT = 9;
	const A_NEGATIVE_WEIGHT = -4;

	test('addBidirectionalEdge', () => {
		const graph = matrixGraph(VERTICES)();

		graph.addBidirectionalEdge(X, Y, A_WEIGHT);

		expect(graph.getCost(X, Y)).toBe(A_WEIGHT);
		expect(graph.getCost(Y, X)).toBe(A_WEIGHT);
	});

	test('addEdgeAndGetCost', () => {
		const graph = matrixGraph(VERTICES)();

		graph.addEdge(X, Y, A_WEIGHT);

		expect(graph.getCost(X, Y)).toBe(A_WEIGHT);

		graph.addEdge(X, Y, A_NEGATIVE_WEIGHT);

		expect(graph.getCost(X, Y)).toBe(A_NEGATIVE_WEIGHT);
	});

	test('hasNegativeEdges', () => {
		const graph = matrixGraph(VERTICES)();

		graph.addEdge(X, Y, A_WEIGHT);

		expect(graph.hasNegativeEdges()).toBe(false);

		graph.addEdge(X, Y, A_NEGATIVE_WEIGHT);

		expect(graph.hasNegativeEdges()).toBe(true);
	});

	test('haveConnection', () => {
		const graph = matrixGraph(VERTICES)();

		graph.addEdge(X, Y, A_WEIGHT);

		expect(graph.haveConnection(X, Y)).toBe(true);

		graph.removeEdge(X, Y);

		expect(graph.haveConnection(X, Y)).toBe(false);
	});

	test('neighboursOf', () => {
		const graph = matrixGraph(VERTICES)();

		graph.addEdge(X, Y, A_WEIGHT);
		graph.addEdge(X, Z, A_WEIGHT);
		graph.addEdge(Y, W, A_WEIGHT);

		const expected: Edge[] = [
			{destination: Y, cost: A_WEIGHT},
			{destination: Z, cost: A_WEIGHT}
		];
		expect(graph.neighboursOf(X)).toBe(expected);
	});

	test('removeEdge', () => {
		const graph = matrixGraph(VERTICES)();

		graph.addBidirectionalEdge(X, Y, A_WEIGHT);

		expect(graph.haveConnection(X, Y)).toBe(true);
		expect(graph.haveConnection(Y, X)).toBe(true);

		graph.removeEdge(X, Y);

		expect(graph.haveConnection(X, Y)).toBe(false);
		expect(graph.haveConnection(Y, X)).toBe(false);
	});

	test('vertices', () => {
		const graph = matrixGraph(VERTICES)();

		expect(graph.vertices()).toBe(VERTICES);
	});
});
